java-checks/src/main/java/org/sonar/java/checks/aws/AwsRegionShouldBeSetExplicitlyCheck.java

```diff
@@ -0,0 +1,217 @@
+        return;
+      }
+      // If the usage sets the region, we return
+      if (usages.stream().anyMatch(AwsRegionShouldBeSetExplicitlyCheck::setsRegion)) {
+        return;
+      }
+      reportIssue(actualDeclaration, "Set the region explicitly on this builder.");
+    } else {
+      reportIssue(invocation, "Set the region explicitly on this builder.");
+    }
+  }
+
+  /**
+   * Crawls up a call chain using methodSelect to determine whether one of the methods called sets the region.
+   *
+   * @param invocation The first in the chain to inspect
+   * @return True if the region is potentially set by one of the method calls in the chain
+   */
+  private static boolean setsRegion(MethodInvocationTree invocation) {
+    CallChainVisitor visitor = new CallChainVisitor(REGION_METHOD);
+    invocation.accept(visitor);
+    return visitor.methodIsInvoked;
+  }
+
+  /**
+   * Crawls down the method calls following an SdkBuilder identifier to check if one of them returns sets the region.
+   *
+   * @param identifier The SdkBuilder to check
+   * @return True if one of the method calls sets the region. False otherwise
+   */
+  private static boolean setsRegion(IdentifierTree identifier) {
+    ReverseCallChainVisitor visitor = new ReverseCallChainVisitor(REGION_METHOD);
+    identifier.accept(visitor);
+    return visitor.methodIsInvoked;
+  }
+
+  private static Optional<VariableTree> getDeclaration(MethodInvocationTree terminalCall) {
+    ExpressionTree expression = terminalCall.methodSelect();
+    while (expression.is(Tree.Kind.MEMBER_SELECT)) {
+      MemberSelectExpressionTree memberSelectExpressionTree = (MemberSelectExpressionTree) expression;
+      ExpressionTree currentExpression = memberSelectExpressionTree.expression();
+      if (currentExpression.is(Tree.Kind.IDENTIFIER)) {
+        IdentifierTree identifier = (IdentifierTree) currentExpression;
+        Tree declaration = identifier.symbol().declaration();
+        if (declaration != null && declaration.is(Tree.Kind.VARIABLE)) {
+          return Optional.of((VariableTree) declaration);
+        }
+      }
+      if (!currentExpression.is(Tree.Kind.METHOD_INVOCATION)) {
+        return Optional.empty();
+      }
+      MethodInvocationTree currentInvocation = (MethodInvocationTree) currentExpression;
+      expression = currentInvocation.methodSelect();
+    }
+    return Optional.empty();
+  }
+
+  private static boolean isPassedAsArgument(IdentifierTree identifier) {
+    Tree parent = identifier.parent();
+    return parent != null && parent.is(Tree.Kind.ARGUMENTS);
+  }
+
+  private static class CallChainVisitor extends BaseTreeVisitor {
+    protected boolean methodIsInvoked;
+    private MethodMatchers target;
+
+    CallChainVisitor(MethodMatchers matcher) {
+      this.target = matcher;
+    }
+
+    @Override
+    public void visitMethodInvocation(MethodInvocationTree tree) {
```

Comment:
This kind of code looks like an excellent candidate for generalization. Since you're going to do the ticket about setting the credentials provider next, I assume you already plan to generalize it then anyway

Reply:
Indeed that is part of what can be easily extracted